src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```
@@ -1,25 +1,79 @@
 package com.es.phoneshop.model.product;
 
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Currency;
 import java.util.List;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+    private List<Product> products;
+    private long currId;
+    private final Object lock = new Object();
+
+    public ArrayListProductDao() {
+        this.products = new ArrayList<>();
+        saveSampleProducts();
+    }
+
     @Override
-    public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+    public Product getProduct(Long id) throws ProductNotFoundException {
+        synchronized (lock) {
+            return products.stream().
+                    filter(product -> id.equals(product.getId()))
+                    .findAny()
+                    .orElseThrow(ProductNotFoundException::new);
+        }
     }
 
     @Override
     public List<Product> findProducts() {
-        throw new RuntimeException("Not implemented");
+        synchronized (lock) {
+            return products.stream()
+                    .filter(product -> product.getPrice() != null)
+                    .filter(product -> product.getStock() >0)
+                    .collect(Collectors.toList());
+        }
     }
 
     @Override
     public void save(Product product) {
-        throw new RuntimeException("Not implemented");
+        synchronized (lock) {
+            if (product.getId() != null) {
+                delete(product.getId());
+            }
+            product.setId(currId++);
+            products.add(product);
+        }
     }
 
     @Override
     public void delete(Long id) {
-        throw new RuntimeException("Not implemented");
+        synchronized (lock) {
+            products = products.stream()
+                    .filter(currProduct -> !id.equals(currProduct.getId()))
```

There is a suitable method called **removeIf**. Try to use it.
Moreover your implementation will silently skip the case when we pass non existing product id but it would be better to throw exception